File: schemas/video-conferencing/drizzle/meeting_participants.ts
// meeting_participants: Tracks each user's participation in a meeting with media state.
// See README.md for full design rationale.

import { pgTable, pgEnum, uuid, text, boolean, integer, timestamp, index, unique } from "drizzle-orm/pg-core";
import { meetings } from "./meetings";
import { users } from "./users";

export const participantRoleEnum = pgEnum("participant_role", ["host", "co_host", "moderator", "attendee", "viewer"]);

export const meetingParticipants = pgTable(
  "meeting_participants",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    meetingId: uuid("meeting_id").notNull().references(() => meetings.id, { onDelete: "cascade" }),
    userId: uuid("user_id").references(() => users.id, { onDelete: "set null" }),
    displayName: text("display_name").notNull(),
    role: participantRoleEnum("role").notNull().default("attendee"),
    joinedAt: timestamp("joined_at", { withTimezone: true }).notNull().defaultNow(),
    leftAt: timestamp("left_at", { withTimezone: true }),
    durationSeconds: integer("duration_seconds"),
    isCameraOn: boolean("is_camera_on").notNull().default(false),
    isMicOn: boolean("is_mic_on").notNull().default(false),
    isScreenSharing: boolean("is_screen_sharing").notNull().default(false),
    createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
    updatedAt: timestamp("updated_at", { withTimezone: true }).notNull().defaultNow(),
  },
  (table) => [
    unique("uq_meeting_participants_meeting_id_user_id").on(table.meetingId, table.userId),
    index("idx_meeting_participants_user_id").on(table.userId),
    index("idx_meeting_participants_meeting_id_joined_at").on(table.meetingId, table.joinedAt),
  ]
);

File: schemas/video-conferencing/spacetimedb/meeting_participants.rs
// meeting_participants: users who have joined a meeting session.
// See README.md for full design rationale.

use spacetimedb::{SpacetimeType, Timestamp};

#[derive(SpacetimeType, Clone)]
pub enum ParticipantRole {
    Host, // type: String
    CoHost,
    Moderator,
    Attendee,
    Viewer,
}

// Composite unique (meeting_id, user_id) — enforce in reducer logic

#[spacetimedb::table(name = meeting_participants, public)]
pub struct MeetingParticipant {
    #[primary_key]
    pub id: String, // UUID
    #[index(btree)]
    pub meeting_id: String, // UUID — FK → meetings.id (cascade delete)
    #[index(btree)]
    pub user_id: Option<String>, // UUID — FK → users.id (set null on delete)
    pub display_name: String,
    pub participant_role: ParticipantRole,
    pub joined_at: Timestamp,
    pub left_at: Option<Timestamp>,
    pub duration_seconds: Option<i32>,
    pub is_camera_on: bool,
    pub is_mic_on: bool,
    pub is_screen_sharing: bool,
    pub created_at: Timestamp,
    pub updated_at: Timestamp,
}
